synapse/handlers/room_batch.py

```diff
@@ -281,8 +282,13 @@ async def persist_historical_events(
             room_id: Room where you want the events persisted in.
             inherited_depth: The depth to create the events at (you will
                 probably by calling inherit_depth_from_prev_ids(...)).
-            auth_event_ids: Define which events allow you to create the given
-                event in the room.
+            state_event_ids:
+                The full state at a given event. We share the same state across the whole
+                historical batch. The state events will be stripped down to only what's
+                necessary to auth a given event and set as the auth_event_ids. For
+                insertion events, we will add all of these state events as the explicit
+                state so the rest of the historical batch can inherit the same state and
+                state_group.
```

Comment:
Ah. We probably don't actually want this behaviour, instead we want to only set the state for the insertion event and allow subsequent events to inherit the state from their prev event. When just sending messages this doesn't matter, but if we send state events in the batch then the current behaviour means that the state event won't become part of the state.

For example, if a user joins half way through the batch and they won't be able to send the message (assuming they weren't added as part of `state_events_at_start`).

I think its trivial to change, that we only set `state_event_ids` for the initial insertion event, i.e. when the prev_events are empty.

Or am I forgetting how this works yet again?

Reply:
> instead we want to only set the state for the insertion event and allow subsequent events to inherit the state from their prev event.

This is what currently is happening. And as far as I can tell, the comment explains it that way too.

For non-insertion events, we only use `state_event_ids` to derive `auth_event_ids` to set.

For insertion events, we set the state as `state_event_ids`

> For example, if a user joins half way through the batch and they won't be able to send the message (assuming they weren't added as part of `state_events_at_start`).

It's not expected to have any state events in the batch of `events`. Any messages being sent should have `m.room.member` already in the state for that point in time where we are inserting at or added to `state_events_at_start`.

Reply:
I really don't know why I get confused about this *every time*. Ah well.

From the doc-comment:

> We share the same state across the whole historical batch.

Ok, so I think is what is getting me. I think what you mean is that the state for the historical batch is all derived from the state at the insertion event? My reading of it was "all events in the batch have the same state", which is not true?


> For non-insertion events, we only use `state_event_ids` to derive `auth_event_ids` to set.

Why do we need this for non-insertion events? Can't we just let the code derive it as normal from the state? That way the only special case is handling the insertion event.



> It's not expected to have any state events in the batch of `events` (or at least processed to auth further messages). Any messages being sent should have `m.room.member` already in the state for that point in time where we are inserting at or added to `state_events_at_start`.

This means that the timeline will never see when a user joins/leaves, which may be fine for the Gitter use case, but we probably want to support it more generally than other use cases. I think we get this for free though, if we stop specifying the auth events directly.

Reply:
> Why do we need this for non-insertion events? Can't we just let the code derive it as normal from the state? That way the only special case is handling the insertion event.

Ahh, this does work and is a good simplification! Updated comments to be more nuanced here too ⏩

I also tried doing something similar for the `state_events_at_start` chain to avoid the copies but since each state event is an `outlier` which uses `EventContext.for_outlier()` and has no `state_ids` to derive from, we have to explicitly set the state for each and can't avoid it as far as I can tell. I also commented the code with these details.

Overall, more understanding of why things are the way they are from this prompt to change 👍



> This means that the timeline will never see when a user joins/leaves, which may be fine for the Gitter use case, but we probably want to support it more generally than other use cases. I think we get this for free though, if we stop specifying the auth events directly.

The problem is that the events are persisted in reverse-chronological order so they have the correct `stream_ordering` while being backfilled which decrements.

As an example of why this fails, if we `/batch_send` -> `events: [messageFromA, memberEventB, messageFromB]`, it will first try to persist `messageFromB` which will fail because `memberEventB` hasn't been processed yet.

Reply:
> The problem is that the events are persisted in reverse-chronological order so they have the correct `stream_ordering` while being backfilled which decrements.
> 
> As an example of why this fails, if we `/batch_send` -> `events: [messageFromA, memberEventB, messageFromB]`, it will first try to persist `messageFromB` which will fail because `memberEventB` hasn't been processed yet.

Oh yes, nyargh. I think we want to fix that, but not in this PR I guess. My hunch is that we want to try and persist the entire batch of events all in one go, rather than doing them one-by-one in reverse order.

Reply:
For my own reference, here is a Complement test around adding state in the middle of the `events` batch (it fails for reason outlined above).

<details>
<summary>Complement test for "State in the middle of the events batch resolves for subsequent events"</summary>

```go
t.Run("State in the middle of the events batch resolves for subsequent events", func(t *testing.T) {
	t.Parallel()

	roomID := as.CreateRoom(t, createPublicRoomOpts)
	alice.JoinRoom(t, roomID, nil)

	// Create the "live" event we are going to import our historical events next to
	eventIDsBefore := createMessagesInRoom(t, alice, roomID, 1, "eventIDsBefore")
	eventIdBefore := eventIDsBefore[0]
	timeAfterEventBefore := time.Now()

	events := createMessageEventsForBatchSendRequest([]string{virtualUserID}, timeAfterEventBefore, 1)

	// Add some events from Ricky who joined in the middle of the events batch
	virtualUserID2 := "@ricky:hs1"
	ensureVirtualUserRegistered(t, as, "ricky")
	events = append(
		events,
		createJoinStateEventsForBatchSendRequest([]string{virtualUserID2}, timeAfterEventBefore)...,
	)
	events = append(
		events,
		createMessageEventsForBatchSendRequest([]string{virtualUserID2}, timeAfterEventBefore, 1)...,
	)

	// Import a historical event
	batchSendHistoricalMessages(
		t,
		as,
		roomID,
		eventIdBefore,
		"",
		createJoinStateEventsForBatchSendRequest([]string{virtualUserID}, timeAfterEventBefore),
		events,
		// Status
		200,
	)
})
```

</details>